src/main/scala/org/broadinstitute/dsde/workbench/leonardo/auth/SamAuthProvider.scala

```diff
@@ -0,0 +1,218 @@
+    }
+  }
+
+  protected def getProjectActionString(action: ProjectAction): String = {
+    action match {
+      case CreateClusters => "launch_notebook_cluster"
+      case ListClusters => "list_notebook_cluster"
+      case SyncDataToClusters => "sync_notebook_cluster"
+      case DeleteClusters => "delete_notebook_cluster"
+      case _ => throw NotebookActionError(action)
+    }
+  }
+
+  protected def getNotebookClusterActionString(action: NotebookClusterAction): String = {
+    action match {
+      case GetClusterStatus => "status"
+      case ConnectToCluster => "connect"
+      case SyncDataToCluster => "sync"
+      case DeleteCluster => "delete"
+      case _ => throw NotebookActionError(action)
+    }
+  }
+
+  /**
+    * @param userEmail The user in question
+    * @param action The project-level action (above) the user is requesting
+    * @param googleProject The Google project to check in
+    * @return If the given user has permissions in this project to perform the specified action.
+    */
+  def hasProjectPermission(userEmail: WorkbenchEmail, action: ProjectActions.ProjectAction, googleProject: String)(implicit executionContext: ExecutionContext): Future[Boolean] = {
+    Future{ resourcesApiAsPet(userEmail, googleProject).resourceAction(billingProjectResourceTypeName, googleProject, getProjectActionString(action)) }
+  }
+
+  /**
+    * Leo calls this method to verify if the user has permission to perform the given action on a specific notebook cluster.
+    * It may call this method passing in a cluster that doesn't exist. Return Future.successful(false) if so.
+    *
+    * @param userEmail      The user in question
+    * @param action        The cluster-level action (above) the user is requesting
+    * @param googleProject The Google project the cluster was created in
+    * @param clusterName   The user-provided name of the Dataproc cluster
+    * @return If the userEmail has permission on this individual notebook cluster to perform this action
+    */
+  def hasNotebookClusterPermission(userEmail: WorkbenchEmail, action: NotebookClusterActions.NotebookClusterAction, googleProject: String, clusterName: String)(implicit executionContext: ExecutionContext): Future[Boolean] = {
+    // get the id for the cluster resource
+    val clusterResourceId = getClusterResourceId(googleProject, clusterName)
+
+    // if action is connect, check only cluster resource. If action is anything else, either cluster or project must be true
+    Future {
+      val notebookAction = resourcesApiAsPet(userEmail, googleProject).resourceAction(notebookClusterResourceTypeName, clusterResourceId, getNotebookClusterActionString(action))
+      if (action == ConnectToCluster) {
+        notebookAction
+      } else {
+        notebookAction || resourcesApiAsPet(userEmail, googleProject).resourceAction(billingProjectResourceTypeName, googleProject, getNotebookClusterActionString(action))
```

Comment:
as i said before you're duplicating what `LeoService` does here. just look at the notebook action here

Reply:
here's what LeoService does:
https://github.com/DataBiosphere/leonardo/blob/develop/src/main/scala/org/broadinstitute/dsde/workbench/leonardo/service/LeonardoService.scala#L198

i.e. it calls both and does the OR itself. we do this specifically to make auth provider implementations simpler to write :)